src/ui/map.js

```diff
@@ -2764,6 +2764,10 @@ class Map extends Camera {
             this.painter._updateFog(this.style);
             this._updateTerrain(); // Terrain DEM source updates here and skips update in style._updateSources.
             this.style._updateSources(this.transform);
+            // Update positions of markers on enabling/disabling terrain
+            for (const marker of this._markers) {
```

Comment:
TBH I don't *immediately* see why this is the correct location at which to update judiciously in order to fix the bug, but it looks reasonable, and if it fixes the issue, then it seems correct.

Reply:
could you put these three lines into ` _forceMarkerUpdate()` function so that its easier to reason about and use in other cases as they come up?

Reply:
@arindam1993 Done!
@rreusser The update needs to happen after `_updateSources` is called, since the markers derive their position from `map.transform` which has no knowledge of terrain until that function is called.